### ISSUE TITLE
单细胞多样本整合之R语言版scVI

### ISSUE THREAD

**Issue by ixxmu** [ixxmu/mp_duty#3705]
https://mp.weixin.qq.com/s/BnkCf_gyjMGirBJ4SunSNQ

**Comment by ixxmu** [comment]
单细胞多样本整合之R语言版scVI by 生信菜鸟团
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p data-tool="mdnice编辑器">上个帖子简单介绍了scVI和scANVI，以及其python环境部署，并尝试运行了一个示例数据，详见：</p><ul data-tool="mdnice编辑器"><li><section><a href="https://mp.weixin.qq.com/s?__biz=Mzg5MjcxNzg1NA==&amp;mid=2247487358&amp;idx=1&amp;sn=a9dff19083623ef3063df6ccce9ee411&amp;scene=21#wechat_redirect" data-linktype="2">单细胞多样本整合之scVI和scANVI</a></section></li></ul><p data-tool="mdnice编辑器">利用reticulate可以在R语言中运行scVI，本期推文对此做一简单介绍。</p><blockquote data-tool="mdnice编辑器"><p>scVI官网教程在https://docs.scvi-tools.org/en/stable/tutorials/notebooks/harmonization.html，github在https://github.com/scverse/scvi-tools</p><p>scVI属于比较高级的整合算法（常见于各大高分文章），<strong>但仍需强调的是，scVI/scANVI整合分析比较吃资源，如果有GPU的话速度会快5-10倍；如果仅用CPU运行scVI，还需谨慎。因为CPU模式的scVI运行速度挺慢的，特别是针对大样本单细胞数据的运行。</strong></p></blockquote><h3 data-tool="mdnice编辑器">一. 环境部署</h3><p data-tool="mdnice编辑器">虽然使用R语言运行，但仍需部署一个scvi的Python环境，这部分与推文【<a href="https://mp.weixin.qq.com/s?__biz=Mzg5MjcxNzg1NA==&amp;mid=2247487358&amp;idx=1&amp;sn=a9dff19083623ef3063df6ccce9ee411&amp;scene=21#wechat_redirect" data-linktype="2">单细胞多样本整合之scVI和scANVI</a>】的环境部署部分一样：</p><pre data-tool="mdnice编辑器"><span></span><code>mamba create -n scvi python=3.9<br>conda activate scvi<br><br><span>#安装软件</span><br>pip install scvi-tools anndata numpy scanpy scib certifi scib-metrics pymde scvi-colab<br><span>#Please be sure to install a version of PyTorch that is compatible with your GPU (if applicable).</span><br></code></pre><p data-tool="mdnice编辑器">还需要在Python里运行：</p><pre data-tool="mdnice编辑器"><span></span><code><span>from</span> scvi_colab <span>import</span> install<br><br>install()<br></code></pre><p data-tool="mdnice编辑器">如果你的服务器有GPU的话，可以安装下面的包进行GPU加速：</p><pre data-tool="mdnice编辑器"><span></span><code>mamba install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia -y<br>mamba install jax jaxlib -c conda-forge -y<br></code></pre><p data-tool="mdnice编辑器">如果下述R包没有安装的，需要在R语言中安装：</p><pre data-tool="mdnice编辑器"><span></span><code><span># install.packages("Seurat")</span><br><span># install.packages("reticulate")</span><br><span># install.packages("cowplot")</span><br><span># install.packages("devtools")</span><br><br><span># devtools::install_github("satijalab/seurat-data")</span><br><span># SeuratData::InstallData("pbmc3k")</span><br><span># install.packages("https://seurat.nygenome.org/src/contrib/ifnb.SeuratData_3.0.0.tar.gz", repos = NULL, type = "source") </span><br><span># SeuratData::InstallData("ifnb")</span><br><br><span># devtools::install_github("cellgeni/sceasy")</span><br></code></pre><h3 data-tool="mdnice编辑器">二. 运行示例数据</h3><h4 data-tool="mdnice编辑器">Step1. 读入数据</h4><p data-tool="mdnice编辑器">示例数据采用经典的<code>ifnb</code>数据集。</p><pre data-tool="mdnice编辑器"><span></span><code><span>## R包加载</span><br><span>library</span>(Seurat)<br><span>library</span>(stringr)<br><span>library</span>(ggplot2)<br><span>library</span>(patchwork)<br><span>library</span>(reticulate)<br><span>library</span>(sceasy)<br><span>library</span>(readr)<br><span>library</span>(cowplot)<br><span>library</span>(SeuratData)<br>sc &lt;- import(<span>"scanpy"</span>, convert = <span>FALSE</span>)<br>scvi &lt;- import(<span>"scvi"</span>, convert = <span>FALSE</span>)<br><br><span>### 1.读入数据</span><br>data(<span>"ifnb"</span>)<br></code></pre><h4 data-tool="mdnice编辑器">Step2. 标准化+高变基因选取</h4><pre data-tool="mdnice编辑器"><span></span><code><span>### 2. 标准分析流程</span><br>seurat.data &lt;- ifnb %&gt;% NormalizeData(normalization.method = <span>"LogNormalize"</span>, <br>                                             scale.factor = <span>10000</span>) %&gt;% <br>  FindVariableFeatures(selection.method = <span>"vst"</span>, nfeatures = <span>2000</span>)<br>top2000 &lt;- head(VariableFeatures(seurat.data), <span>2000</span>)<br>seurat.data &lt;- seurat.data[top2000]<br></code></pre><h4 data-tool="mdnice编辑器">Step3. 将Seurat数据转化为AnnData对象</h4><p data-tool="mdnice编辑器">关于Seurat数据转化为AnnData对象（R语言单细胞与Python数据互转），很久之前我介绍过一个R/Python包<code>diopy</code>，详见：</p><ul data-tool="mdnice编辑器"><li><section><a href="https://mp.weixin.qq.com/s?__biz=Mzg5MjcxNzg1NA==&amp;mid=2247484778&amp;idx=1&amp;sn=45f99192489b835d895bd2bd4d93e774&amp;scene=21#wechat_redirect" data-linktype="2">巧用Python加速单细胞分析</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI0NDA4OTQ2NA==&amp;mid=2652571513&amp;idx=1&amp;sn=878ed662a9a61ebd4e67faf21032f6d4&amp;scene=21#wechat_redirect" data-linktype="2">scDIOR软件打通单细胞组学R/Python平台壁垒</a></section></li></ul><p data-tool="mdnice编辑器">这里想和大家分享另一款R包<code>sceasy</code>，加上<code>outFile</code>参数即可输出<code>h5ad</code>的数据，衔接Python非常方便：</p><pre data-tool="mdnice编辑器"><span></span><code><span>### 3.将seurat转为AnnData</span><br>adata &lt;- convertFormat(seurat.data, <br>                       <span>#outFile = "./python.data/NatCellBiol.Chen.2021.Batch1.h5ad",</span><br>                       from=<span>"seurat"</span>, <br>                       to=<span>"anndata"</span>, <br>                       main_layer=<span>"counts"</span>, <br>                       drop_single_values=<span>FALSE</span>,<br>                       <span>#outFile='filename.h5ad'</span><br>                      )<br>print(adata) <span># Note generally in Python, dataset conventions are obs x var</span><br></code></pre><pre data-tool="mdnice编辑器"><span></span><code><span> AnnData object with n_obs × n_vars = 13997 × 2000</span><br><span>        obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'stim', 'seurat_annotations', 'percent.mt'</span><br><span>        var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'vst.variable'</span><br></code></pre><h4 data-tool="mdnice编辑器">Step4. scVI整合</h4><pre data-tool="mdnice编辑器"><span></span><code><span>### 3.整合</span><br><span># run setup_anndata, use column stim for batch</span><br>scvi$model$SCVI$setup_anndata(adata, batch_key = <span>'stim'</span>)<br><br><span># create the model</span><br>model = scvi$model$SCVI(adata)<br><br><span># train the model</span><br>model$train()<br><br><span># to specify the number of epochs when training:</span><br><span># model$train(max_epochs = as.integer(400))</span><br><br><span># get the latent represenation</span><br>latent = model$get_latent_representation()<br><br><span># put it back in our original Seurat object</span><br>latent &lt;- as.matrix(latent)<br>rownames(latent) = colnames(ifnb)<br>ifnb[[<span>"scvi"</span>]] &lt;- CreateDimReducObject(embeddings = latent,<br>                                       key = <span>"scvi_"</span>, <br>                                       assay = DefaultAssay(ifnb))<br>ifnb &lt;- RunUMAP(ifnb, dims = <span>1</span>:<span>10</span>, reduction = <span>"scvi"</span>, n.components = <span>2</span>)<br></code></pre><p data-tool="mdnice编辑器">这里可以学习一下如何使用<code>CreateDimReducObject</code>函数哦。</p><p data-tool="mdnice编辑器">UMAP可视化：</p><pre data-tool="mdnice编辑器"><span></span><code>p1 &lt;- DimPlot(ifnb, reduction = <span>"umap"</span>, group.by = <span>"stim"</span>, pt.size=<span>2</span>)<br>p1<br></code></pre><figure data-tool="mdnice编辑器"><img data-ratio="0.7852998065764023" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPc5zHic2co8icVqpibfB3e7r01lBRU6VXOEhypEc6wDCicEkbY34fO1NT2Yw/640?wx_fmt=other" data-type="other" data-w="517" src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPc5zHic2co8icVqpibfB3e7r01lBRU6VXOEhypEc6wDCicEkbY34fO1NT2Yw/640?wx_fmt=other"><figcaption>image-20230806160515248</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>FeaturePlot(ifnb, features = c(<span>"SELL"</span>, <span>"CREM"</span>, <span>"CD8A"</span>, <span>"GNLY"</span>, <span>"CD79A"</span>, <span>"FCGR3A"</span>, <br>    <span>"CCL2"</span>, <span>"PPBP"</span>), min.cutoff = <span>"q9"</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-ratio="0.8352788586251622" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPcEjX4P29SfoHpwIk1wXumxib9Zjo6TOuqErA9G0cuAU8OgEjvPZhprXg/640?wx_fmt=other" data-type="other" data-w="771" src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPcEjX4P29SfoHpwIk1wXumxib9Zjo6TOuqErA9G0cuAU8OgEjvPZhprXg/640?wx_fmt=other"><figcaption>image-20230806160549246</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>FeaturePlot(ifnb, features = c(<span>"GNLY"</span>, <span>"IFI6"</span>), split.by = <span>"stim"</span>, max.cutoff = <span>3</span>, <br>    cols = c(<span>"grey"</span>, <span>"red"</span>))<br></code></pre><figure data-tool="mdnice编辑器"><img data-ratio="0.8400520156046815" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPc8tnfiaHPiabQWCAYdqiaRdIH0Kt9e7T5rQYIDJd3VHURA5gynfgkuBa9w/640?wx_fmt=other" data-type="other" data-w="769" src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPc8tnfiaHPiabQWCAYdqiaRdIH0Kt9e7T5rQYIDJd3VHURA5gynfgkuBa9w/640?wx_fmt=other"><figcaption>image-20230806160617758</figcaption></figure><h3 data-tool="mdnice编辑器">三. 番外</h3><p data-tool="mdnice编辑器">总体感受：R语言版本的scVI还是Python版的运行的舒服一些，因此我建议能使用Python的尽量还是使用Python。另外，以后有志于专注单细胞数据分析的，特别是大样本单细胞数据分析，我觉得还是有必要学习一点Python。</p><p data-tool="mdnice编辑器">最后，上期推文【<a href="https://mp.weixin.qq.com/s?__biz=Mzg5MjcxNzg1NA==&amp;mid=2247487358&amp;idx=1&amp;sn=a9dff19083623ef3063df6ccce9ee411&amp;scene=21#wechat_redirect" data-linktype="2">单细胞多样本整合之scVI和scANVI</a> 】我使用CPU跑了好几个小时。前几天我借到一台GPU服务器，然后运行同样的示例数据只需要10分钟：</p><figure data-tool="mdnice编辑器"><img data-ratio="0.3138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPcibrowZdBXHdkGf45cnYiaD2PQIuotbOfaE9r0bGxNbAiaLr988A85Yaww/640?wx_fmt=other" data-type="other" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2LosibKb9QCVgNakFpkrLriaYIPcibrowZdBXHdkGf45cnYiaD2PQIuotbOfaE9r0bGxNbAiaLr988A85Yaww/640?wx_fmt=other"><figcaption>image-20230806175249886</figcaption></figure><p data-tool="mdnice编辑器">希望有一天能够服务器自由，GPU自由~</p><span>- END -</span></section><p><br></p><p><mp-style-type data-value="10000"></mp-style-type></p></div>

**Comment by xiaoshanJi98** [comment]
Hello,

I'm encountering an issue while trying to integrate data using scvi-tools in R. Here is the code I am using:

```
latent <- as.matrix(latent)
rownames(latent) = colnames(adata)
adata[["scvi"]] <- CreateDimReducObject(embeddings = latent, key = "scvi_", assay = DefaultAssay(adata))
```
However, this results in the following error:
Error in UseMethod(generic = "DefaultAssay", object = object): no applicable method for 'DefaultAssay' applied to an object of class "c('anndata._core.anndata.AnnData', 'python.builtin.object')"
Traceback:
1. CreateDimReducObject(embeddings = latent, key = "scvi_", assay = DefaultAssay(adata))
2. DefaultAssay(adata)

It seems that the DefaultAssay function is not recognizing the adata object as expected. I have confirmed that adata is an AnnData object. For context, I am using Seurat version 4.

Could you please help me understand what might be causing this error and how to resolve it?

Thank you!
